lib/services/local/users.go

```diff
@@ -73,6 +75,125 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
 	return trace.Wrap(s.DeleteRange(ctx, startKey, backend.RangeEnd(startKey)))
 }
 
+func (s *IdentityService) listUsersWithSecrets(ctx context.Context, pageSize int, pageToken string) ([]types.User, string, error) {
+	rangeStart := backend.Key(webPrefix, usersPrefix, pageToken)
+	rangeEnd := backend.RangeEnd(backend.ExactKey(webPrefix, usersPrefix))
+
+	// Adjust page size, so it can't be too large.
+	if pageSize <= 0 || pageSize > apidefaults.DefaultChunkSize {
+		pageSize = apidefaults.DefaultChunkSize
+	}
+
+	// Artificially inflate the limit to account for user secrets
+	// which have the same prefix.
+	limit := pageSize * 4
+
+	rangeStream := backend.StreamRange(ctx, s.Backend, rangeStart, rangeEnd, limit)
+
+	var (
+		out      []types.User
+		prevUser string
+		items    userItems
+	)
+	for rangeStream.Next() {
+		item := rangeStream.Item()
+
+		name, suffix, err := splitUsernameAndSuffix(string(item.Key))
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		if prevUser == "" {
+			prevUser = name
+		}
+
+		if name != prevUser {
+			user, err := userFromUserItems(prevUser, items)
+			if err != nil {
+				return nil, "", trace.Wrap(err)
+			}
+
+			out = append(out, user)
+			if len(out) == pageSize {
+				break
+			}
+
+			prevUser = name
+			items = userItems{}
+		}
+
+		items.Set(suffix, item)
+	}
+
+	// If there are no more users its possible that the previous user being processed
+	// was never added to out because there was no additional user to transition to.
+	// Add the user from the collected items and return the full output.
+	if len(out) > 0 && out[len(out)-1].GetName() != prevUser {
```

Comment:
Won't this fail to return anything if there's only a single user? A simpler check would be to see if `items` has anything:

```suggestion
	if items.complete() || items.hasLocalAuthItems() {
```

Maybe we could add an `isEmpty` method to `userItems` and use it here.